### PR TITLE
feat(mdsn): allow directly special characters in dsn

### PR DESCRIPTION
**Pull request by zitsen**
Based on https://taosdata.feishu.cn/wiki/VTSIwIV4OiSFGYky2Q4cUFw2nSg

### CODE REVIEW

File: mdsn/src/lib.rs [hunk 1]
```diff
@@ -93,6 +93,12 @@ pub enum DsnError {
     ParseErr(#[from] pest::error::Error<Rule>),
     #[error("unable to parse port from {0}")]
     PortErr(#[from] ParseIntError),
+    #[error("invalid dsn format: {0}, please use <driver>://<username>:<password>@<addresses>/<database>?<params>")]
+    InvalidFormat(String),
+    #[error("No : found in {0}")]
```
Comment (qevolg) [inline mdsn/src/lib.rs:98]: Wouldn't it be better to use ':'?

File: mdsn/src/lib.rs [hunk 2]
```diff
@@ -263,111 +269,289 @@ pub struct Dsn {
...
+            let mut protocol = None;
+            if let Some(at) = main
+                .bytes()
+                .enumerate()
+                .rev()
+                .find(|(_, v)| *v == b'@')
+                .map(|(at, _)| at)
+            {
+                let (user_pass, addrs) = main.split_at(at);
+                let mut addrs = addrs.strip_prefix('@').expect("strip prefix @");
```
Comment (qevolg) [inline mdsn/src/lib.rs:325]: Wouldn't it be better to return an error instead of expect?

File: mdsn/src/lib.rs (hunk 1 above)
Comment (zitsen) [reply mdsn/src/lib.rs:98]: Good suggestion, got it.

File: mdsn/src/lib.rs (hunk 2 above)
Comment (zitsen) [reply mdsn/src/lib.rs:325]: use `split_at` by '@' position, so `expect` may be better than a never reachable error.

Comment (qevolg) [reply mdsn/src/lib.rs:325]: You are correct, I overlooked the `find` above.

**Pull request merged**